### ISSUE TITLE
Bump Parachain memory limit to 1GB

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot#6557]
In the really early days we had parachains running with a 1GB memory limit. Then later I refactored the code and removed the limit, but left this nice issue here: https://github.com/paritytech/polkadot/issues/699 
But after some refactorings it got closed, without being done. :D 

Currently Parachains are running with 128MB, which is the same that Parachains use for their block import pipeline (when they are using Cumulus). However, Parachains should get more memory for validation than they are using for block import. The reason being that they need to keep more data in the wasm memory like the state proof etc. 

1GB is discuss-able, but also not such a huge amount of memory for validators. 

Could be done as part of: https://github.com/paritytech/polkadot/pull/6161

CC @eskimor

**Comment by eskimor** [comment]
yes, once #6161 is merged we should be able to easily bump this up securely.

**Comment by eskimor** [comment]
@bkchr What more data do they need to keep exactly. State proof is limited by PoV size - right, so we are talking about [3Meg](https://github.com/paritytech/polkadot/blob/405fdeb214ef582c39e93ebbaf3dca536a29f2d7/primitives/src/v4/mod.rs#L348) additional memory tops?

**Comment by ggwpez** [comment]
What does this mean for the minimal memory requirement of validators?  
How many of these can run in parallel at once in the worst case?

**Comment by eskimor** [comment]
> What does this mean for the minimal memory requirement of validators? How many of these can run in parallel at once in the worst case?

That is indeed a concern I am sharing: The higher the limit, the less workers we can have in parallel or the higher the hardware requirements. Hence we should not bump this limit higher than necessary.

**Comment by bkchr** [comment]
> What more data do they need to keep exactly. State proof is limited by PoV size - right, so we are talking about [3Meg](https://github.com/paritytech/polkadot/blob/405fdeb214ef582c39e93ebbaf3dca536a29f2d7/primitives/src/v4/mod.rs#L348) additional memory tops?

We not only need to keep the PoV in memory. We also have all values that are being written in memory etc. More info here: https://github.com/paritytech/srlabs_findings/issues/255

BTW, the PoV size is also measured for the compressed PoV. This means that the unpacked PoV can be much bigger.

We may not need to go to 1GB and can propose a lower limit for now.

**Comment by Sophia-Gold** [comment]
> > What more data do they need to keep exactly. State proof is limited by PoV size - right, so we are talking about [3Meg](https://github.com/paritytech/polkadot/blob/405fdeb214ef582c39e93ebbaf3dca536a29f2d7/primitives/src/v4/mod.rs#L348) additional memory tops?
> 
> We not only need to keep the PoV in memory. We also have all values that are being written in memory etc. More info here: [paritytech/srlabs_findings#255](https://github.com/paritytech/srlabs_findings/issues/255)
> 
> BTW, the PoV size is also measured for the compressed PoV. This means that the unpacked PoV can be much bigger.
> 
> We may not need to go to 1GB and can propose a lower limit for now.

Can you comment on what mitigations are planned of those suggested by srlabs? It seems we could at least do 512mb and that is based on the decoding overhead and nested storage transactions so maybe it's possible we could go lower if those things are improved. 

If this affects the hardware requirements it really changes the tone of the governance proposal...

**Comment by bkchr** [comment]
> If this affects the hardware requirements it really changes the tone of the governance proposal...

The standard hardware that we are using to run the benchmarking is currently using 16GB of main memory. This standard hardware is currently not really being chosen with Parachains in mind, besides maybe the network interface. 16GB in 2023 is really not that much and an upgrade to 32GB is really a no brainer IMO. But even with 16GB and having 4 parallel validation workers would lead to 2GB of memory usage, which still gives the node 14GB for other things.

**Comment by eskimor** [comment]
512 does indeed sound quite reasonable already. I am not really in the loop of main memory costs in data centers, but if requiring 32GB barely adds costs these days, worth considering ramping up min requirements, future proofing a bit for adding more leeway for more workers, although then we should also ramp up the minimum core requirement (currently [only 4](https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot)). Anyhow, probably the wrong place to discuss this.

**Comment by eskimor** [comment]
Always on Kusama first.

**Comment by ggwpez** [comment]
Validators are advised to use at least 32GiB ram currently. Looking at the telemetry for nodes shows 39% using at least 64GiB. Would be nice if we could get this just for validators, but it looks generally fine.

Is this just a value in the host config? Then you can easily propose it to Kusama OpenGov on the Fellowship whitelisted track.

**Comment by Sophia-Gold** [comment]
> Is this just a value in the host config? Then you can easily propose it to Kusama OpenGov on the Fellowship whitelisted track.

Yes, that's the plan.
